### PR TITLE
Add config for BERT

### PR DESCRIPTION
**Pull request by mhagiwara**

### CODE REVIEW

File: quick_start/my_text_classifier_bert.jsonnet
Comment (mhagiwara) [inline quick_start/my_text_classifier_bert.jsonnet:1]: @matt-gardner This is the BERT config files I'm going to use for the "next steps" chapter. There's this thing called `bert_pooler` (a `Seq2VecEncoder `) which I've never used before and am not sure how it fits with the rest of the pipeline.

Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet:1]: As in, I know you have to pool the embedding for the `[CLS]` token but what should we use as the TokenEmbedder if we used BertPooler as a Seq2VecEncoder?

Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet:1]: It was my misunderstanding. I just realized you simply need to combine BertEmbedder and BertPooler. But as it is, you need to modify both the embedder and the encoder to support BERT, and add some cryptic fields (`allow_unmatched_keys` and `embedder_to_indexer_map`). The point of the "next steps" chapter is to show how easy it is to swap in a different contextualizer but this could have an opposite effect. I'm starting to wonder if this should make it to the Part 2 at all. @matt-gardner

Comment (matt-gardner) [reply quick_start/my_text_classifier_bert.jsonnet:1]: Today I'm working on removing the need for those cryptic fields (https://github.com/allenai/allennlp/issues/3304).  The other parts of that should be expected - I need to modify the tokenizer, token indexer, and whatever parts of my model used non-BERT stuff.

We can also note that we could have written the code to be specific to BERT / transformers in the first place, requiring very little change in a config file to switch from BERT to RoBERTa or whatever, though that would mean we're stuck with transformers.  You can trade off that flexibility if you want, still getting all of the benefits of our training loop, data processing code, etc., while having your code be specific to transformers.

Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet:1]: So you think it's ok to show this config and briefly (only briefly) talk about how readers should go about using pretrained contextualizers? I'm talking about here https://github.com/allenai/allennlp-course/blame/master/chapters/next-steps.md#L21

Comment (matt-gardner) [reply quick_start/my_text_classifier_bert.jsonnet:1]: I think so - it's just changing a few things (all of the parts you would expect) in the config.  When the crazy parameters are removed, it should be a pretty simple thing to show, with a pointer to a more in-depth discussion elsewhere.

Comment (matt-gardner) [reply quick_start/my_text_classifier_bert.jsonnet:1]: Do you disagree?

Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet:1]: No, I think we can show this. "How to use BERT" is one of the most commonly asked questions about AllenNLP, so we can be upfront about this.

Do you think this PR is good to merge? I also added a requirements file for binder. When this gets merged we can point to this repo so that binder can do imports properly.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (matt-gardner) [inline quick_start/my_text_classifier_bert.jsonnet]: We removed this; you must be using an old version of allennlp.  The tokenizer to use now is the `pretrained-transformer` one.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (matt-gardner) [inline quick_start/my_text_classifier_bert.jsonnet]: This should be a `pretrained-transformer` indexer.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (matt-gardner) [inline quick_start/my_text_classifier_bert.jsonnet]: I finished https://github.com/allenai/allennlp/pull/3597 yesterday; once that's merged, this can go away.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (matt-gardner) [inline quick_start/my_text_classifier_bert.jsonnet]: This also should be `pretrained-transformer`.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (matt-gardner) [inline quick_start/my_text_classifier_bert.jsonnet:31]: This *should* work with the other `pretrained-transformer` stuff that I recommended, though I haven't explicitly tested that.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (matt-gardner) [inline quick_start/my_text_classifier_bert.jsonnet]: Is this the right optimizer for a bert model?  I know this is just an example, but we should probably have the right recommendations even in the example.

**Review by matt-gardner — COMMENTED** [review]
Sorry for being slow to get back to this; I had surgery on Thursday, which put me out of commission for a few days.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet]: OK, I'm only using the latest *released* version of AllenNLP (0.9.0). To clarify, all the code and course content should be compliant with AllenNLP 1.0.0 correct?

File: quick_start/my_text_classifier_bert.jsonnet
Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet]: OK, I'm putting this PR on hold until that gets merged

File: quick_start/my_text_classifier_bert.jsonnet
Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet]: I'm not sure if there's a "right" optimizer for a bert model (adam worked just fine for one of my past BERT projects) but the recommendation in the community, as far as I understand, is using bert_adam with a warmup using a low learning rate (e.g., 1.0e-5). I'll try this configuration and make sure the metrics are at least decent.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet]: Also there was a quite large change in how tokenizers work https://github.com/allenai/allennlp/pull/3361 so we need to change some code in this example repo as well as the course code

File: quick_start/my_text_classifier_bert.jsonnet
Comment (mhagiwara) [reply quick_start/my_text_classifier_bert.jsonnet]: This worked, although it doens't have a paramter `requires_grad` which is necessary for fixing the bert parameters. Do you know if this is being worked on, or am I missing something?

Comment (matt-gardner) [reply quick_start/my_text_classifier_bert.jsonnet]: Why do you want to fix the bert parameters?  We should add that parameter, but you typically don't want to fix the parameters when using bert.

File: binder/requirements.txt
Comment (mhagiwara) [inline binder/requirements.txt]: @matt-gardner Do you think we should install from the latest master instead, e.g., `git+https://github.com/allenai/allennlp.git`? Is there a good commit / tag that we should anchor this to?

Comment (matt-gardner) [reply binder/requirements.txt]: For now, yeah, latest master is probably best.  We don't have a specific tag, and won't until we release 1.0.  I'd probably just grab the SHA from the latest commit and use that, updating it when there's something that we want to incorporate here.

**Comment by mhagiwara** [comment]
@matt-gardner I think this is read for review!

File: quick_start/my_text_classifier_bert.jsonnet
Comment (matt-gardner) [reply quick_start/my_text_classifier_bert.jsonnet]: I'd just remove these two keys now.  That PR is going to be merged in the next hour.

File: quick_start/my_text_classifier_bert.jsonnet
Comment (matt-gardner) [inline quick_start/my_text_classifier_bert.jsonnet:37]: Just a heads up, we're probably going to be removing the `BasicIterator`.  Probably we'll make it so you don't have to specify a type here, and we can safely remove this key.  It's fine with me to leave this here for now, this is just a note that this will probably have to change before the course is released.

**Review by matt-gardner — APPROVED** [review]

**Comment by mhagiwara** [comment]
@matt-gardner the latest commit incorporates your change that makes the embedder key mapping unnecessary. I confirmed this works. If this looks good to you I'll go ahead and merge this

**Comment by matt-gardner** [comment]
LGTM!

**Pull request merged**
